icu4c/source/tools/genrb/filterrb.h

```diff
@@ -0,0 +1,88 @@
+// © 2018 and later: Unicode, Inc. and others.
+// License & terms of use: http://www.unicode.org/copyright.html
+
+#ifndef __FILTERRB_H__
+#define __FILTERRB_H__
+
+#include <list>
+#include <string>
+#include <map>
+#include <ostream>
+
+#include "unicode/utypes.h"
+
+
+class ResKeyPath {
+  public:
+    /** Constructs an empty path (top of tree) */
+    ResKeyPath();
+
+    /** Constructs from a string path */
+    ResKeyPath(const std::string& path, UErrorCode& status);
+
+    void push(const std::string& key);
+    void pop();
+
+    const std::list<std::string>& pieces() const;
+
+  private:
+    std::list<std::string> fPath;
+
+    friend std::ostream& operator<<(std::ostream& out, const ResKeyPath& value);
+};
+
+std::ostream& operator<<(std::ostream& out, const ResKeyPath& value);
+
+
+class PathFilter {
+  public:
+    enum EInclusion {
+        INCLUDE,
+        PARTIAL,
+        EXCLUDE
+    };
+
+    static const char* kEInclusionNames[];
+
+    /**
+     * Returns an EInclusion on whether or not the given path should be included.
+     *
+     * INCLUDE = include the whole subtree
+     * PARTIAL = recurse into the subtree
+     * EXCLUDE = reject the whole subtree
+     */
+    virtual EInclusion match(const ResKeyPath& path) const = 0;
+};
+
+
+class SimpleRuleBasedPathFilter : public PathFilter {
+  public:
+    void addRule(const std::string& ruleLine, UErrorCode& status);
+    void addRule(const ResKeyPath& path, bool inclusionRule, UErrorCode& status);
+
+    EInclusion match(const ResKeyPath& path) const override;
+
+  private:
+    struct Tree {
+        /**
+         * Information on the USER-SPECIFIED inclusion/exclusion.
+         *
+         * INCLUDE = this path exactly matches a "+" rule
+         * PARTIAL = this path does not match any rule, but subpaths exist
+         * EXCLUDE = this path exactly matches a "-" rule
+         */
+        EInclusion fIncluded = PARTIAL;
+        std::map<std::string, Tree> fChildren;
+
+        void print(std::ostream& out, int32_t indent) const;
```

Comment:
I would make a public print function like this in the outer class and call that from operator<< so that you don't have to friend it.

Reply:
Done